File: src/main/java/eu/codetopic/utils/ui/activity/fragment/BaseFragmentActivity.java
package eu.codetopic.utils.ui.activity.fragment;

import android.content.Context;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.IdRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.NavigationView;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.ContextThemeWrapper;
import android.view.View;

import eu.codetopic.java.utils.Objects;
import eu.codetopic.java.utils.log.Log;
import eu.codetopic.utils.AndroidUtils;
import eu.codetopic.utils.R;

public abstract class BaseFragmentActivity extends AppCompatActivity {// TODO: 12.5.16 rework fragment replacing

    @IdRes public static final int CONTAINER_LAYOUT_ID = R.id.base_content;
    public static final String CURRENT_FRAGMENT_TAG =
            "BaseFragmentActivity.CURRENT_FRAGMENT";
    private static final String LOG_TAG = "BaseFragmentActivity";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (savedInstanceState == null) replaceFragment(onCreateMainFragment());
    }

    @Nullable
    private <T extends Fragment> T initFragment(@NonNull Class<T> fragmentClass, @Nullable Bundle bundle) {
        try {
            T fragment = fragmentClass.newInstance();
            if (bundle != null) fragment.setArguments(bundle);
            return fragment;
        } catch (Exception e) {
            Log.e(LOG_TAG, "initFragment", e);
        }
        return null;
    }

    public Fragment getCurrentFragment() {
        return getSupportFragmentManager().findFragmentByTag(CURRENT_FRAGMENT_TAG);
    }

    public void removeCurrentFragment() {
        replaceFragment((Fragment) null);
    }

    public <T extends Fragment> T replaceFragment(@NonNull Class<T> fragmentClass) {
        return replaceFragment(fragmentClass, null);
    }

    public <T extends Fragment> T replaceFragment(@NonNull Class<T> fragmentClass, @Nullable Bundle bundle) {
        return replaceFragment(initFragment(fragmentClass, bundle));
    }

    public <T extends Fragment> T replaceFragment(@NonNull FragmentTransaction ft,
                                                  @NonNull Class<T> fragmentClass, @Nullable Bundle bundle) {
        return replaceFragment(ft, initFragment(fragmentClass, bundle));
    }

    public <T extends Fragment> T replaceFragment(@Nullable T fragment) {
        FragmentTransaction ft = getSupportFragmentManager().beginTransaction();
        try {
            return replaceFragment(ft, fragment);
        } finally {
            ft.commitAllowingStateLoss();
        }
    }

    public <T extends Fragment> T replaceFragment(@NonNull FragmentTransaction ft, @Nullable T fragment) {
        onBeforeReplaceFragment(ft, fragment);
        Fragment currentFragment = getCurrentFragment();

        if (fragment != null) {
            if (currentFragment == null
                    || !Objects.equals(fragment.getClass(), currentFragment.getClass())
                    || !AndroidUtils.equalBundles(fragment.getArguments(), currentFragment.getArguments())) {
                ft.replace(CONTAINER_LAYOUT_ID, fragment, CURRENT_FRAGMENT_TAG);
            }
            return fragment;
        }

        if (currentFragment != null) ft.remove(currentFragment);
        setTitle(AndroidUtils.getApplicationLabel(this));
        return null;
    }

    protected void onBeforeReplaceFragment(FragmentTransaction ft, Fragment fragment) {
    }

    protected abstract Fragment onCreateMainFragment();

    @Override
    public void onAttachFragment(Fragment fragment) {
        super.onAttachFragment(fragment);
        if (CURRENT_FRAGMENT_TAG.equals(fragment.getTag())) {
            if (fragment instanceof TitleProvider) {
                setTitle(((TitleProvider) fragment).getTitle());
            }
            if (fragment instanceof ThemeProvider) {
                int themeId = ((ThemeProvider) fragment).getThemeId();
                Context themedContext = new ContextThemeWrapper(getLayoutInflater().getContext(), themeId);
                Toolbar toolbar = findViewById(R.id.toolbar);
                if (toolbar != null) {

                    int backgroundColor = AndroidUtils.getColorFromAttr(
                            themedContext, R.attr.colorPrimary, -1);
                    if (backgroundColor != -1) toolbar.setBackgroundColor(backgroundColor);

                    int titleTextColor = AndroidUtils.getColorFromAttr(
                            themedContext, R.attr.titleTextColor, -1);
                    if (titleTextColor != -1) toolbar.setTitleTextColor(titleTextColor);

                    int subtitleTextColor = AndroidUtils.getColorFromAttr(
                            themedContext, R.attr.subtitleTextColor, -1);
                    if (subtitleTextColor != -1) toolbar.setSubtitleTextColor(subtitleTextColor);

                    if (Build.VERSION.SDK_INT >= 21) {
                        int navigationBarColor = AndroidUtils.getColorFromAttr(
                                themedContext, android.R.attr.navigationBarColor, -1);
                        if (navigationBarColor != -1)
                            getWindow().setNavigationBarColor(navigationBarColor);

                        int statusBarColor = AndroidUtils.getColorFromAttr(
                                themedContext, android.R.attr.statusBarColor, -1);
                        if (statusBarColor != -1)
                            getWindow().setStatusBarColor(statusBarColor);
                    }
                }

                NavigationView navView = findViewById(R.id.nav_view);
                if (navView != null) {
                    View headerView = navView.getHeaderView(0).findViewById(R.id.navigationHeader);
                    if (headerView != null) {
                        int headerBackgroundColor = AndroidUtils.getColorFromAttr(
                                themedContext, R.attr.colorPrimaryDark, -1);
                        if (headerBackgroundColor != -1)
                            headerView.setBackgroundColor(headerBackgroundColor);
                    }
                }
            }
        }

        //System.runFinalization();
        //System.gc();
    }
}

File: src/main/java/eu/codetopic/utils/ui/activity/fragment/BaseFragmentActivity.kt
/*
 * utils
 * Copyright (C)   2017  anty
 *
 * This program is free  software: you can redistribute it and/or modify
 * it under the terms  of the GNU General Public License as published by
 * the Free Software  Foundation, either version 3 of the License, or
 * (at your option) any  later version.
 *
 * This program is distributed in the hope that it  will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied  warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
 * GNU General Public License for more details.
 *
 * You  should have received a copy of the GNU General Public License
 * along  with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package eu.codetopic.utils.ui.activity.fragment

import android.annotation.SuppressLint
import android.app.ActivityManager
import android.os.Build
import android.os.Bundle
import android.support.annotation.IdRes
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentTransaction
import android.support.v7.app.AppCompatActivity
import android.view.ContextThemeWrapper
import android.view.View

import eu.codetopic.java.utils.letIf
import eu.codetopic.java.utils.log.Log
import eu.codetopic.utils.AndroidUtils
import eu.codetopic.utils.use
import eu.codetopic.utils.R
import eu.codetopic.utils.ui.view.hideKeyboard
import kotlinx.android.extensions.CacheImplementation
import kotlinx.android.extensions.ContainerOptions
import kotlinx.android.synthetic.main.activity_module_toolbar.*
import kotlinx.android.synthetic.main.activity_navigation_base.*

@ContainerOptions(CacheImplementation.SPARSE_ARRAY)
abstract class BaseFragmentActivity : AppCompatActivity() {// TODO: 12.5.16 rework fragment replacing

    companion object {

        private const val LOG_TAG = "BaseFragmentActivity"

        const val FRAGMENT_TAG_CURRENT = "BaseFragmentActivity.CURRENT_FRAGMENT"

        @IdRes val LAYOUT_ID_CONTAINER = R.id.base_content
    }

    val currentFragment: Fragment?
        get() = supportFragmentManager.findFragmentByTag(FRAGMENT_TAG_CURRENT)

    protected abstract val mainFragmentClass: Class<out Fragment>?

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        title = getDefaultTitle()

        if (savedInstanceState == null) replaceFragment(onCreateMainFragment())
    }

    override fun onContentChanged() {
        super.onContentChanged()
        updateProviderFragment()
    }

    override fun onStart() {
        super.onStart()
        updateProviderFragment()
    }

    private fun <T : Fragment> initFragment(fragmentClass: Class<out T>, bundle: Bundle? = null): T? = try {
        fragmentClass.newInstance().apply {
            bundle?.let { arguments = bundle }
        }
    } catch (e: Exception) { Log.e(LOG_TAG, "initFragment", e); null }

    fun removeCurrentFragment() {
        replaceFragment(null as Fragment?)
    }

    @JvmOverloads
    fun <T : Fragment> replaceFragment(fragmentClass: Class<out T>, bundle: Bundle? = null): T? =
            replaceFragment(initFragment(fragmentClass, bundle))

    @JvmOverloads
    fun <T : Fragment> replaceFragment(ft: FragmentTransaction,
                                       fragmentClass: Class<out T>, bundle: Bundle? = null): T? =
            replaceFragment(ft, initFragment(fragmentClass, bundle))

    fun <T : Fragment> replaceFragment(fragment: T?): T? {
        val ft = supportFragmentManager.beginTransaction()
        return try {
            try {
                return replaceFragment(ft, fragment)
            } finally {
                ft.commit()
            }
        } catch (e: Exception) {
            Log.e(LOG_TAG, "replaceFragment($fragment) -> Failed to replace fragment", e); null
        }
    }

    fun <T : Fragment> replaceFragment(ft: FragmentTransaction, fragment: T?): T? {
        onBeforeReplaceFragment(ft, fragment)

        currentFocus?.hideKeyboard()

        val currFragment = currentFragment

        return fragment?.apply {
            takeIf {
                currFragment == null
                        || it.javaClass != currFragment.javaClass
                        || !AndroidUtils.equalBundles(it.arguments, currFragment.arguments)
            }?.let { ft.replace(LAYOUT_ID_CONTAINER, it, FRAGMENT_TAG_CURRENT) }
        } ?: run {
            currFragment?.let {
                ft.remove(currFragment)
                title = getDefaultTitle()
            }
            null
        }
    }

    protected open fun onBeforeReplaceFragment(ft: FragmentTransaction, fragment: Fragment?) {}

    protected open fun onCreateMainFragment(): Fragment? = try {
        mainFragmentClass?.newInstance()
    } catch (e: Exception) {
        Log.e(LOG_TAG, "onCreateMainFragment() -> Failed to create main fragment", e)
        null
    }

    override fun onAttachFragment(fragment: Fragment) {
        super.onAttachFragment(fragment)

        if (FRAGMENT_TAG_CURRENT == fragment.tag)
            applyProviderFragment(fragment)

        //System.runFinalization();
        //System.gc();
    }

    protected open fun getDefaultTitle(): CharSequence = AndroidUtils.getAppLabel(this)

    fun updateProviderFragment() = applyProviderFragment(currentFragment)

    @SuppressLint("InlinedApi")
    private fun applyProviderFragment(fragment: Fragment?) {
        val targetTitle = if (fragment is TitleProvider) fragment.title else getDefaultTitle()

        val themedContext =
                if (fragment is ThemeProvider)
                    ContextThemeWrapper(layoutInflater.context, fragment.themeId)
                else this

        val colorsAttrs = intArrayOf(
                R.attr.colorPrimary,
                R.attr.titleTextColor,
                R.attr.subtitleTextColor,
                R.attr.colorPrimaryDark
        ).letIf({ Build.VERSION.SDK_INT >= 21 }) {
            it + intArrayOf(
                    android.R.attr.navigationBarColor,
                    android.R.attr.statusBarColor
            )
        }

        val colors = themedContext.obtainStyledAttributes(colorsAttrs)
                .use { tArray ->
                    (0 until colorsAttrs.size).map {
                        tArray.getColor(it, -1)
                                .takeIf { it != -1 }
                    }
                }

        title = targetTitle

        if (Build.VERSION.SDK_INT >= 21) {
            run taskDescription@ {
                val primaryColor = colors[colorsAttrs.indexOf(R.attr.colorPrimary)]
                        ?: return@taskDescription
                /*val icon =
                        if (fragment is IconProvider) fragment.icon
                        else AndroidUtils.drawableToBitmap(
                                AndroidUtils.getActivityIcon(themedContext, componentName)
                        )*/
                //val title = targetTitle.toString()

                //setTaskDescription(ActivityManager.TaskDescription(title, icon, primaryColor))
                setTaskDescription(ActivityManager.TaskDescription(null, null, primaryColor))
            }

            colors[colorsAttrs.indexOf(android.R.attr.navigationBarColor)]
                    ?.let { window.navigationBarColor = it }

            colors[colorsAttrs.indexOf(android.R.attr.statusBarColor)]
                    ?.let { window.statusBarColor = it }
        }

        toolbar?.apply {
            colors[colorsAttrs.indexOf(R.attr.colorPrimary)]
                    ?.let { setBackgroundColor(it) }

            colors[colorsAttrs.indexOf(R.attr.titleTextColor)]
                    ?.let { setTitleTextColor(it) }

            colors[colorsAttrs.indexOf(R.attr.subtitleTextColor)]
                    ?.let { setSubtitleTextColor(it) }
        }

        navigationView?.apply {
            getHeaderView(0).findViewById<View>(R.id.boxHeader)?.apply {
                colors[colorsAttrs.indexOf(R.attr.colorPrimaryDark)]
                        ?.let { setBackgroundColor(it) }
            }
        }

    }
}